documentstore_migracao/utils/convert_html_body.py

```diff
@@ -943,6 +944,73 @@ def transform(self, data):
             _process(xml, "graphic", self.parser_node)
             return data
 
+    class RemoveReferencesFromBodyPipe(CustomPipe):
+        def transform(self, data):
+            raw, xml = data
+            if len(self.super_obj.ref_items) > 0:
+                p_references = self._mark_references(xml)
+
+                self._mark_the_last_reference_end(xml)
+
+                p_items = self._find_reference_paragraphs(xml)
+
+                if len(self.super_obj.ref_items) == len(p_items):
+                    self._add_p_references_to_delete(p_references, p_items)
+                    self._remove_references(xml, p_items)
+                else:
+                    logger.info("RemoveReferencesFromBodyPipe: FALHOU")
+            return data
+
+        def _mark_references(self, xml):
+            comments = xml.xpath("//comment()")
+            p_references = None
+            for comment in comments:
+                name = comment.text.strip()
+                if name == "ref":
+                    if p_references is None:
+                        p_references = comment.getprevious()
+                    mark = etree.Element("REF-TO-DELETE")
+                    comment.addnext(mark)
+            return p_references
+
+        def _add_p_references_to_delete(self, p_references, p_items):
+            if p_references is not None and p_references.tag == "p":
+                text = get_node_text(p_references)
+                if text.lower().startswith("ref"):
+                    logger.info(
+                        "RemoveReferencesFromBodyPipe: Primeiro: %s" % text
+                    )
+                    p_items.insert(0, p_references)
+
+        def _remove_references(self, xml, p_to_delete):
+            for p in p_to_delete:
+                parent = p.getparent()
+                if parent is not None:
+                    parent.remove(p)
+            etree.strip_tags(xml, "REF-TO-DELETE")
+            etree.strip_tags(xml, "REF-FIM")
+
+        def _mark_the_last_reference_end(self, xml):
+            ref_items = xml.findall(".//REF-TO-DELETE")
+            if len(ref_items) > 0:
+                p = ref_items[-1].getnext()
+                if p is not None:
+                    mark = etree.Element("REF-FIM")
+                    p.addnext(mark)
+
+        def _find_reference_paragraphs(self, xml):
+            p_items = []
+            e = xml.find(".//REF-TO-DELETE")
+            if e is not None:
+                while True:
+                    e = e.getnext()
+                    if e is None or e.tag == "REF-FIM":
+                        break
+                    if e.tag == "p":
+                        if get_node_text(e).strip():
+                            p_items.append(e)
+            return p_items
```

Comment:
Levando em consideração o comentário anterior eu também alterei esse método para poder capturar as referências a partir das tags auxiliares.

```python
        def _find_reference_paragraphs(self, xml):
            """Captura todos os parágrafos contendo referências a partir
            da tag auxiliar `REF-TO-DELETE`"""
            paragraphs = []
            references_end = xml.findall(".//REF-TO-DELETE")

            for element in references_end:
                parent = element.getparent()

                if not get_node_text(parent).strip():
                    continue

                paragraphs.append(parent)

            return paragraphs
```

Levando em consideração que no passo anterior foi possível detectar todas as referências então nós não precisaríamos mais da tag `REF-FIM` já que o `xml.findall(".//REF-TO-DELETE")` nos trará todos os nós que precisam ser deletados.